### ISSUE TITLE
crash in compiling M2 IntegralClosure.  FLINT.

### ISSUE THREAD

**Issue by kschwede** [Macaulay2/M2#1964]
Hi all,

I have been trying to compile M2 (*not* with cmake).  WSL2, Ubuntu 20.04, master branch.

Everything looks like it's going fine but when I get to:
`making example results for "integralClosure(...,Strategy=>...)`
it errors out.

```
i78 : time R' = integralClosure(R, Strategy => Radical)
Exception (FLINT memory_manager). Unable to allocate memory (8).
Aborted (core dumped)
```
The full error log of that example is attached here:

[_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.zip](https://github.com/Macaulay2/M2/files/6045463/_integral__Closure_lp._cm__Strategy_eq_gt._rp.zip)

I seem to have plenty of memory.

```
schwede@2021AMD:~/M2Source/M2/M2$ free
              total        used        free      shared  buff/cache   available
Mem:       52580800      172452    45414068          72     6994280    51791028
Swap:      13631488           0    13631488
```

**Comment by d-torrance** [comment]
This looks like it might be related to #1423.  We switched to a smaller example to throw a band-aid on the problem, but there's probably still an underlying memory leak.

**Comment by DanGrayson** [comment]
By the way, we purposely limit memory usage during the running of tests and examples, so you can't expect it to make use of all of your memory.  You could play around with the memory limits to see how close that example is to succeeding.  In run.m2, the memory limits are set in the line
```
ulimit := utest "-c unlimited" | utest "-t 700" | utest "-m 850000" | utest "-s 8192" | utest "-n 512"
```
and in
```
GCMAXHEAP      = 1 << 21 -* sets GC_MAXIMUM_HEAP_SIZE=400M *-
```

**Comment by kschwede** [comment]
I tried

```
ulimit := utest "-c unlimited" | utest "-t 7000" | utest "-m 8500000" | utest "-s 81920" | utest "-n 5120"
```
and
```
GCMAXHEAP      = 1 << 21 -* sets GC_MAXIMUM_HEAP_SIZE=4000M *-
```

But it still crashed in the same way...

**Comment by DanGrayson** [comment]
On the GCMAXHEAP line you edited to comment instead of the code.  Try changing the number 21 to something larger.

**Comment by kschwede** [comment]
Sorry for the delay, I changed it from 21 to 210, same error...

However, I see:

```
 ulimit -c unlimited; ulimit -t 7000; ulimit -m 8500000; ulimit -s 81920;  cd /tmp/M2-15063-0/4-rundir/; GC_MAXIMUM_HEAP_SIZE=400M "/home/schwede/M2Source/M2/M2/usr-dist/x86_64-Linux-Ubuntu-20.04/bin/M2-binary" -q --int --no-randomize --no-readline --silent --stop --print-width 77 -e 'needsPackage("IntegralClosure",Reload=>true,FileName=>"/home/schwede/M2Source/M2/M2/Macaulay2/packages/IntegralClosure.m2")' <"/tmp/M2-15063-0/0_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.m2" >>"/home/schwede/M2Source/M2/M2/usr-dist/common/share/doc/Macaulay2/IntegralClosure/example-output/_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.errors" 2>&1
/home/schwede/M2Source/M2/M2/usr-dist/common/share/doc/Macaulay2/IntegralClosure/example-output/_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.errors:0:1: (output file) error: Macaulay2 killed by signal 6
/tmp/M2-15063-0/0_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.m2:0:1: (input file)
/tmp/M2-15063-0/4-rundir/: error: files remain in temporary run directory after program exits abnormally
```


so maybe I am changing it incorrectly?  The relevant line of run.m2 says:

```
GCMAXHEAP      = 1 << 210 -* sets GC_MAXIMUM_HEAP_SIZE=400M *-
```

**Comment by moorewf** [comment]
I think that is a bitshift operator, so you may want to try a number such as 23 or 24 instead of 210.

What does the `.errors` file say?

**Comment by DanGrayson** [comment]
Yes, a number like 1<<210, when converted a machine int, will come out as 0.

**Comment by mahrud** [comment]
Revert everything else, this is the line you want to edit:
https://github.com/Macaulay2/M2/blob/f0ce581750dfd94336883695d3552671410f32f5/M2/Macaulay2/m2/run.m2#L97

**Comment by kschwede** [comment]
I tried it with:
```
GCMAXHEAP      = 1 << 23 -* sets GC_MAXIMUM_HEAP_SIZE=400M *-
```
still the same error.

I'm including the .errors file.
[_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.zip](https://github.com/Macaulay2/M2/files/6096265/_integral__Closure_lp._cm__Strategy_eq_gt._rp.zip)

**Comment by kschwede** [comment]
> > On the GCMAXHEAP line you edited to comment instead of the code. Try changing the number 21 to something larger.
> 
> That's wrong. This is the line you want to edit:
> https://github.com/Macaulay2/M2/blob/f0ce581750dfd94336883695d3552671410f32f5/M2/Macaulay2/m2/run.m2#L97
> 
> 
> The bit shifts are just a way to signal which arguments to include and which to ignore.

I tried changing it to 4000M, didn't recompile from scratch, just ran `make` in place again.  Same error.

```
ulimit -c unlimited; ulimit -t 700; ulimit -m 850000; ulimit -s 8192; ulimit -n 512;  cd /tmp/M2-28464-0/1-rundir/; GC_MAXIMUM_HEAP_SIZE=4000M "/home/schwede/M2Source/M2/M2/usr-dist/x86_64-Linux-Ubuntu-20.04/bin/M2-binary" -q --int --no-randomize --no-readline --silent --stop --print-width 77 -e 'needsPackage("IntegralClosure",Reload=>true,FileName=>"/home/schwede/M2Source/M2/M2/Macaulay2/packages/IntegralClosure.m2")' <"/tmp/M2-28464-0/0_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.m2" >>"/home/schwede/M2Source/M2/M2/usr-dist/common/share/doc/Macaulay2/IntegralClosure/example-output/_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.errors" 2>&1
/home/schwede/M2Source/M2/M2/usr-dist/common/share/doc/Macaulay2/IntegralClosure/example-output/_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.errors:0:1: (output file) error: Macaulay2 killed by signal 6
/tmp/M2-28464-0/0_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.m2:0:1: (input file)
/tmp/M2-28464-0/1-rundir/: error: files remain in temporary run directory after program exits abnormally
M2: *** Error 6
 -- 1.96258 seconds elapsed
```

**Comment by mahrud** [comment]
Does running this also fail?
```
/home/schwede/M2Source/M2/M2/usr-dist/x86_64-Linux-Ubuntu-20.04/bin/M2-binary -q --int --no-randomize --no-readline --silent --stop --print-width 77 -e 'needsPackage("IntegralClosure",Reload=>true,FileName=>"/home/schwede/M2Source/M2/M2/Macaulay2/packages/IntegralClosure.m2")' <"/tmp/M2-28464-0/0_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.m2"
```

**Comment by kschwede** [comment]
> Does running this also fail?
> 
> ```
> /home/schwede/M2Source/M2/M2/usr-dist/x86_64-Linux-Ubuntu-20.04/bin/M2-binary -q --int --no-randomize --no-readline --silent --stop --print-width 77 -e 'needsPackage("IntegralClosure",Reload=>true,FileName=>"/home/schwede/M2Source/M2/M2/Macaulay2/packages/IntegralClosure.m2")' <"/tmp/M2-28464-0/0_integral__Closure_lp..._cm__Strategy_eq_gt..._rp.m2"
> ```

When I tried running that, I just got a prompt
```
>
```
with no way to exit (besides ctrl-c).

However, if I run the sequence of commands that leads to that error:

```
S' = QQ[a..f];
M' = genericSymmetricMatrix(S',a,3);
I' = minors(2,M');
center = ideal(b,e,a-d,d-f);
S = QQ[a,b,d,e];
p = map(S'/I',S,gens center);
I = kernel p;
R = S/I;
time R' = integralClosure(R, Strategy => Radical)
```

inside the executable `/home/schwede/M2Source/M2/M2/usr-dist/x86_64-Linux-Ubuntu-20.04/bin/M2-binary`
then I still crash.  

```
i19 : time R' = integralClosure(R, Strategy => Radical)
Exception (FLINT memory_manager). Unable to allocate memory (8).
Aborted
```

Running the same commands in something I compiled a few weeks does *not* crash.

Has something in FLINT changed?

**Comment by DanGrayson** [comment]
Now I'm getting the same problem trying to build 1.18 under Ubuntu 20.04, and it is blocking the release.
```m2
i78 : time R' = integralClosure(R, Strategy => Radical)
Exception (FLINT memory_manager). Unable to allocate memory (8).
Aborted (core dumped)
```

**Comment by d-torrance** [comment]
> Now I'm getting the same problem trying to build 1.18 under Ubuntu 20.04, and it is blocking the release.
> 
> ```m2
> i78 : time R' = integralClosure(R, Strategy => Radical)
> Exception (FLINT memory_manager). Unable to allocate memory (8).
> Aborted (core dumped)
> ```

One possible idea -- maybe try downgrading flint to 2.6.3?  That's the version that's currently in Debian unstable/testing and the most recent releases of Ubuntu, and I've had no issues like this with it.

Unfortunately, it's not available in Ubuntu 20.04 as a package in the official repositories -- that still has 2.5.2, which didn't have the `factor` improvements IIRC.

**Comment by DanGrayson** [comment]
Thanks for the suggestion.  I think first I'll try debugging...

**Comment by DanGrayson** [comment]
Debugging suggests that it might be a bug in factory 4.1.1, which is provided by Ubuntu 20.04, and which calls flint with bad arguments.  So I'm enabling the building of factory, which will build 4.2.0.

...

Yes, that fixes it.  I'll change the configure script so it's not satisfied with 4.1.1.

**Comment by DanGrayson** [comment]
Should be resolved by commit 29459064dbc127672059c797adf115f5c326d731, closing.
